examples/tictactoe.rs

```diff
@@ -0,0 +1,297 @@
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		write!(f, "{:?}", self)
+	}
+}
+
+impl From<InterpreterError> for Error {
+	fn from(e: InterpreterError) -> Self {
+		Error::Interpreter(e)
+	}
+}
+
+impl From<DeserializationError> for Error {
+	fn from(e: DeserializationError) -> Error {
+		Error::Deserialize(e)
+	}
+}
+
+impl From<ValidationError> for Error {
+	fn from(e: ValidationError) -> Error {
+		Error::Validation(e)
+	}
+}
+
+impl HostError for Error {}
+
+mod tictactoe {
+	use super::Error;
+
+	#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+	pub enum Player {
+		X,
+		O,
+	}
+
+	#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+	pub enum GameResult {
+		Draw,
+		Won(Player),
+	}
+
+	impl Player {
+		pub fn into_i32(maybe_player: Option<Player>) -> i32 {
+			match maybe_player {
+				None => 0,
+				Some(Player::X) => 1,
+				Some(Player::O) => 2,
+			}
+		}
+	}
+
+	#[derive(Debug)]
+	pub struct Game {
+		board: [Option<Player>; 9],
+	}
+
+	impl Game {
+		pub fn new() -> Game {
+			Game {
+				board: [None; 9],
+			}
+		}
+
+		pub fn set(&mut self, idx: i32, player: Player) -> Result<(), Error> {
+			if idx < 0 || idx > 9 {
+				return Err(Error::OutOfRange);
+			}
+			if self.board[idx as usize] != None {
+				return Err(Error::AlreadyOccupied);
+			}
+			self.board[idx as usize] = Some(player);
+			Ok(())
+		}
+
+		pub fn get(&self, idx: i32) -> Result<Option<Player>, Error> {
+			if idx < 0 || idx > 9 {
+				return Err(Error::OutOfRange);
+			}
+			Ok(self.board[idx as usize])
+		}
+
+		pub fn game_result(&self) -> Option<GameResult> {
+			// 0, 1, 2
+			// 3, 4, 5
+			// 6, 7, 8
+			let patterns = &[
+				// Rows
+				(0, 1, 2),
+				(3, 4, 5),
+				(6, 7, 8),
+
+				// Columns
+				(0, 3, 6),
+				(1, 4, 7),
+				(2, 5, 8),
+
+				// Diagonals
+				(0, 4, 8),
+				(2, 4, 6),
+			];
+
+			// Returns Some(player) if all cells contain same Player.
+			let all_same = |i1: usize, i2: usize, i3: usize| -> Option<Player> {
+				if self.board[i1].is_none() {
+					return None;
+				}
+				if self.board[i1] == self.board[i2] && self.board[i2] == self.board[i3] {
+					return self.board[i1];
+				}
+				None
+			};
+
+			for &(i1, i2, i3) in patterns {
+				if let Some(player) = all_same(i1, i2, i3) {
+					return Some(GameResult::Won(player));
+				}
+			}
+
+			// Ok, there is no winner. Check if it's draw.
+			let all_occupied = self.board.iter().all(|&cell| cell.is_some());
+			if all_occupied {
+				Some(GameResult::Draw)
+			} else {
+				// Nah, there are still empty cells left.
+				None
+			}
+		}
+	}
+}
+
+struct Runtime<'a> {
+	player: tictactoe::Player,
+	game: &'a mut tictactoe::Game,
+}
+
+const SET_FUNC_INDEX: usize = 0;
+const GET_FUNC_INDEX: usize = 1;
+
+impl<'a> Externals for Runtime<'a> {
+	fn invoke_index(
+		&mut self,
+		index: usize,
+		args: &[RuntimeValue],
+	) -> Result<Option<RuntimeValue>, InterpreterError> {
+		match index {
+			SET_FUNC_INDEX => {
+				let idx: i32 = args[0].try_into().unwrap();
```

Comment:
Does the `InterpreterError` cover the "wrong arguments" case? is there an example of how to fire that error?

Reply:
It is a good question.

Usually it is not a problem, because this function *must* be called with arguments specified [at resolve time](https://github.com/paritytech/parity-wasm/pull/128/files/d7b33f91305f53325875e54e9fd9e796fcb0c89d#diff-61d355bade5e256b6b4ab9f9ed333910R195) (unless we haven't mess up interpreter implementation).

However, there is a such thing as polymorphic functions which can be imported, and thus called, with different set of arguments. Because this sounds to me like something host-specific magic, I thought that this is ok to model with `HostErrors`: [related example](https://github.com/paritytech/parity-wasm/pull/128/files/d7b33f91305f53325875e54e9fd9e796fcb0c89d#diff-3ef0eb37cb7ae9c4b7d49c7f0bf53583R137)